First steps to make protocols scheduler agnostic

t-reents opened the pull request:
Fixes #1010 

The current protocols implicitly rely on the `Slurm` scheduler, thus causing errors when users try to use ad different one, e.g. SGE as described in the issue. 
Here, the documentation of the `get_builder_from_protocol` method is extended to make the users aware of this behavior. Moreover, a warning is added in case no explicit `resources` are provided via the `options` argument, as this will cause errors in future versions.

@sphuber @mbercx Suggestions for a different phrasing or place where to put the additional documentation are very welcome. Moreover, I used a `UserWarning` because a `DeprecationWarning` somehow didn't appear. I'm not sure whether they are filtered at some point?

sphuber commented:
> @sphuber @mbercx Suggestions for a different phrasing or place where to put the additional documentation are very welcome.

It would be great to also add this to the documentation in addition to the docstrings. I noticed that the `get_builder_from_protocol` functionality isn't really documented though. 

> Moreover, I used a `UserWarning` because a `DeprecationWarning` somehow didn't appear. I'm not sure whether they are filtered at some point?

Deprecation warnings are ignored by default in Python (https://docs.python.org/3/library/exceptions.html#DeprecationWarning). A user has to enable them manually to see them. I think a `UserWarning` is fine for now as an alternative.

t-reents commented:
Hi @sphuber ! Sorry for picking this up just now.

I added a general description of the `get_builder_from_protocol` in the `How to - Workflows` section. To be honest, I wasn't 100% sure where to put it, but since it's a general remark about the workflows, I decided to put it there. Happy to discuss/adjust this.

Moreover, I realized that the issue of the protocols depending on SLURM also applies to the other workflows, not just the `pw` ones. Hence, I also added the warning to those.

mbercx commented:
Thanks @t-reents, sorry for only looking at this PR now. A few preliminary comments, also after reading the corresponding issue:

1. If we don't remove the Slurm scheduler defaults from the protocol files, trying to use a code on an SGE-scheduler computer in e.g. the `PwRelaxWorkChain.get_builder_from_protocol` method will still fail outright, even if the correct SGE resources are specified through the options, because the `num_machines` input is still there.
2. Removing both these lines:

   https://github.com/aiidateam/aiida-quantumespresso/blob/435fc9008d012c9a0d81d8fad95624dde7b670cb/src/aiida_quantumespresso/workflows/protocols/pw/base.yaml#L13-L14

   will result in a failure in case the `resources` dictionary is not provided by the user. However, this should be a perfectly legitimate use case: for the `direct` scheduler specifying the resources is not really necessary. I think you can also set defaults for Slurm when setting up/configuring the computer (have to double-check, don't have Slurm locally).
3. I think a custom warning is not necessary. If we simply set the `resources` option to an empty dictionary in the protocol YAML files (removing the `num_machines` line), the scheduler inputs will still be validated before the calculation is run. _When_ this will happen depends if you are running the `PwBaseWorkChain` or a high-level work chain like the `PwRelaxWorkChain`. The former will only fail upon submission, but the latter will fail at this line:
   https://github.com/aiidateam/aiida-quantumespresso/blob/435fc9008d012c9a0d81d8fad95624dde7b670cb/src/aiida_quantumespresso/workflows/pw/relax.py#L160
   Because the validation does seem to happen at this point. Either way, a warning that checks the options doesn't seem necessary, because in case the user doesn't specify the correct resources the calculation will not run anyways?

t-reents commented:
Thanks @mbercx for your comments!

I agree that simply removing the `num_machines` would basically fix it and the validation would do the rest of the job. I think that I also mentioned this as a second option in the issue. However, the idea of the warnings is just to have them for an intermediate period, to make the users aware that in one of the next releases, we'll expect explicitly defined resources (btw, I agree with you that it should be fine for the `local` scheduler). 

To be honest, I don't have a strong opinion on this, so if I understand your comments correctly, we can also directly adjust the protocol files and only add the information to the documentation, without introducing these warnings.

mbercx commented:
> To be honest, I don't have a strong opinion on this, so if I understand your comments correctly, we can also directly adjust the protocol files and only add the information to the documentation, without introducing these warnings.

I think this should be the best way forward, considering that the warnings are sometimes unnecessary, and in case they are, the validation would stop the calculation from running anyways? Maybe I'm missing something though.

mbercx commented:
@t-reents I see now we aren't removing the `num_machines` line from the protocol YAML files anymore. Unfortunately, that means that trying to use the protocol methods with e.g. the SGE scheduler will still fail with the following error even when you provide the correct resources in the options:

```python
ValueError: invalid attribute value input `metadata.options.resources` is not valid for the `SgeScheduler` scheduler: these parameters were not recognized: num_machines
```

So if we want to enable SGE users to actually use the protocol methods, I don't think we can get around removing the `num_machines:` line. Unfortunately, if we remove it, users will have to specify `num_machines` in the options for the direct and Slurm schedulers.

One option would be to follow [@sphuber's suggestion](https://github.com/aiidateam/aiida-quantumespresso/issues/1010#issuecomment-1947459351) and simply accept that users will have to specify the resources. I agree with him that implementing scheduler-specific logic in the protocol methods to provide some defaults is not really desirable. However, we'd be breaking backwards-compatibility in a sense, and make it a bit more difficult for users of these schedulers to run calculations. Especially for the direct scheduler this bugs me, why do I have to specify the `num_machines` if I'm just starting to use the plugin package on my own computer?

I think the real solution is that we should be able to specify scheduler defaults easily on the configured computer. But that will of course take quite a bit of time. As a stopgap, I'd be fine with:

1. Removing the `num_machines:` lines from the protocols, so we're not breaking the functionality for SGE users (and potentially others).
2. Adding some **temporary** scheduler-specific code for the direct and Slurm schedulers where necessary, e.g.:

```
        if code.computer.scheduler_type in ['core.slurm', 'core.direct']:
            inputs['pw']['metadata']['options']['resources'].setdefault('num_machines', 1)
```

Let me know if you agree. I'm happy to make the changes and add the necessary tests etc.

t-reents commented:
Hi @mbercx, yeah, I agree.

I'm very busy right now, so please feel free to continue, in case you have the time (If not, I'll do it at some point, but might take some time). I'm of course available to discuss

mbercx commented:
@t-reents I've updated the PR so default resources are added for most of the schedulers in `aiida-core`. Even though I don't think this workaround that adds scheduler-specific code to `aiida-quantumespresso` is the optimal solution, I think it's our best option in case we want to (1) allow for schedulers that don't use `NodeNumberJobResource` and (2) avoid requiring users to specify the default resources.

I've added the `add_option` `staticmethod` so all the logic regarding this workaround is in one place and doesn't need to be duplicated. Disadvantage is that someone looking at the code will have to check this method to see what is happening, and so it obscures the fact that default resources are set a bit. Maybe `set_options` is also a better name for the method? Let me know what you think, and I'll add some tests for the other schedulers if we agree on the approach and implementation.

mbercx commented:
@JPchico maybe give this branch a spin too if you'd like. Comments on the SGE defaults are welcome, since I've never used this scheduler myself.
